### PR TITLE
EOS-24274 : Implement message bus producer with transaction

### PR DESCRIPTION
**Pull request by akash2144**
Added transactional wrapper over kafka producer

Signed-off-by: Akash Dudhane <akash.a.dudhane@seagate.com>

# Problem Statement
- Implement message bus producer with transaction

# Design
-  Message bus implements transactional producer using transaction feature of Kakfa.
    https://www.confluent.io/blog/transactions-apache-kafka/
- Current implementation of Message Bus in utils have guaranteed-delivery of messages but if broker fails it may write duplicates of the retried message in the stream. Transactional producer avoid this problem also it guaranties exactly once delivery of messages.

# Coding
-  [x] Coding conventions are followed and code is consistent

# Testing 
- [x] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Review Checklist 
- [x] PR is self reviewed
- [x] JIRA number/GitHub Issue added to PR
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained
- [ ] Is there a change in filename/package/module or signature? [Y/N]: 
- [ ] If yes for above point, is a notification sent to all other cortx components? [Y/N]
- [ ] Side effects on other features (deployment/upgrade)? [Y/N]
- [ ] Dependencies on other component(s)? [Y/N]
-     If yes for above point, post link to the corresponding PR.

# Review Checklist 
- [ ] Is perfline test run and the report with and without the changes updated in the PR? [Y/N]: 

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

## Unit Test Details
<pre>
  <code>
    <h3>Tested in K8S environment
    <h3> 1. transactional_producer.py
    <h4> 
import sys, os
import pathlib
import time, json
from ha.core.event_manager.event_manager import EventManager
from ha.util.message_bus import MessageBus
from ha.util.msg_bus.tcp.kafka.const import TOPIC

sys.path.append(os.path.join(os.path.dirname(pathlib.Path(__file__)), '..', '..', '..', '..'))
event_manager = EventManager.get_instance()

transactional_producer = MessageBus.get_transactional_producer(producer_id='ha_event_manager_hare', message_type='ha_event_hare')
for x in range(5):
    transactional_producer.transactional_publish(message=f"Message {x}", TOPIC='ha_event_hare')
    print("Message sent.....{x}")

[root@cortx-control-headless-svc /]# python3 test/transactional_producer.py
Hosts :  kafka.default.svc.cluster.local:9092
Message sent.....0
Message sent.....1
Message sent.....2
Message sent.....3
Message sent.....4

     <h3> 2. transactional_consumer.py
     <h4>    
import os
import sys
import pathlib
import time, json
from ha.core.event_manager.event_manager import EventManager
from ha.util.message_bus import MessageBus
from ha.util.msg_bus.tcp.kafka.const import TOPIC

sys.path.append(os.path.join(os.path.dirname(pathlib.Path(__file__)), '..', '..', '..', '..'))
event_manager = EventManager.get_instance()

def process_message(message: str):
        try:
            print("Message Recieved : ", message)
            return 'success'
        except Exception as err:
            print(f'Failed to analyze the event: {err}')

transactional_consumer = MessageBus.get_transactional_consumer(consumer_id='test', consumer_group='test_publisher',
                                                               message_type='ha_event_hare', callback=process_message)
transactional_consumer.start()
while True:
    time.sleep(2)

[root@cortx-control-headless-svc /]# python3 test/transactional_consumer.py
Hosts :  kafka.default.svc.cluster.local:9092
Message Recieved :  Message 0
Message Recieved :  Message 1
Message Recieved :  Message 2
Message Recieved :  Message 3
Message Recieved :  Message 4

  </code>
</pre>

### CODE REVIEW

File: ha/util/msg_bus/message_bus.py [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+
+from cortx.utils.conf_store import Conf
+from ha.util.msg_bus.tcp.kafka import const
+from ha.util.msg_bus.error import InvalidConfigError, ConnectionEstError
+from ha.util.msg_bus.tcp.kafka.kafka import KafkaProducerComm, KafkaConsumerComm
+from cortx.utils.log import Log
+
+_DELIM = ">"
```
Comment (ArchanaLimaye) [inline ha/util/msg_bus/message_bus.py:24]: DELIM is already defined in const

File: ha/util/message_bus.py [hunk 2]
```diff
@@ -142,6 +143,112 @@ def stop(self):
...
+                else:
+                    retry = True
+                    continue
+                retry = False
+            except Exception as e:
+                Log.error(f"Supressing exception from message bus {e}")
+                retry = False
+
+    def start(self):
+        self.consumer = MessageBusComm(comm_type=self.comm_type, group_id=self.group_id,
```
Comment (Madhura-08) [inline ha/util/message_bus.py:242]: From line 157, Same class instantiation returns producer and here, it returns consumer. How is that possible?

File: ha/util/msg_bus/message_bus.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+           a consumer group. For comm_type as CONSUMER this field is required.
+        """
+        #TODO: Add one more field for taking configuration path as a parameter.
+
+        self._comm_obj = None
+        self._message_bus_type = 'kafka'
+        self._comm_type = kwargs.get(const.COMM_TYPE, None)
+        self._init_config(**kwargs)
+
+    def _init_config(self, **kwargs):
```
Comment (Madhura-08) [inline ha/util/msg_bus/message_bus.py:58]: Please add method docstring here and wherever applicable

**Review by Madhura-08 — COMMENTED** [review]
Have you tried with multiple partitions? If yes, can we try following:
- drop a message in between and check whether msg gets discarded from all the partitions?

File: ha/util/message_bus.py (hunk 2 above)
Comment (akash2144) [reply ha/util/message_bus.py:242]: Depending on comm_type it returns producer OR consumer. On line 157 we are passing comm_type as PRODUCER & here comm_type='CONSUMER'

**Pull request closed without merging**